src/backend/input_builder.rs

```
@@ -133,24 +134,58 @@ enum Error {
-// https://support.cloudflare.com/hc/en-us/articles/115001635128-Configuring-Cloudflare-Rate-Limiting
-fn ip_key(ip_str: &str) -> Result<String, Error> {
-    let ip = ip_str.parse::<IpAddr>()?;
-    Ok(match ip {
+/// Generate a string key for backend. Uses more memory but can easily be combined with other
+/// data like path or custom keys.
+///
+/// Groups IPv6 addresses together, see:
+/// https://adam-p.ca/blog/2022/02/ipv6-rate-limiting/
+/// https://support.cloudflare.com/hc/en-us/articles/115001635128-Configuring-Cloudflare-Rate-Limiting
+pub fn string_ip_key(ip_str: Option<&str>) -> String {
+    let ip = parse_ip(ip_str);
+    match ip {
         IpAddr::V4(v4) => v4.to_string(),
         IpAddr::V6(v6) => {
             if let Some(v4) = v6.to_ipv4() {
-                return Ok(v4.to_string());
+                return v4.to_string();
             }
             let zeroes = [0u16; 4];
             let concat = [&v6.segments()[0..4], &zeroes].concat();
             let concat: [u16; 8] = concat.try_into().unwrap();
             let subnet = Ipv6Addr::from(concat);
             format!("{}/64", subnet)
         }
-    })
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum MyIpAddr {
+    V4(Ipv4Addr),
+    V6([u16; 4]),
+}
+
+impl From<IpAddr> for MyIpAddr {
+    fn from(value: IpAddr) -> Self {
+        match value {
+            IpAddr::V4(addr) => MyIpAddr::V4(addr),
+            IpAddr::V6(addr) => MyIpAddr::V6(addr.segments()[..4].try_into().unwrap()),
+        }
+    }
+}
+
+/// Generate a raw byte key for backend which uses less memory.
+pub fn raw_ip_key(ip_str: Option<&str>) -> MyIpAddr {
+    parse_ip(ip_str).into()
+}
+
+fn parse_ip(addr: Option<&str>) -> IpAddr {
+    if let Some(addr) = addr {
+        if let Ok(ip) = IpAddr::from_str(addr) {
+            return ip;
+        } else if let Ok(socket) = SocketAddr::from_str(addr) {
+            return socket.ip();
+        }
+    }
+    Ipv4Addr::new(127, 0, 0, 1).into()
```

the reason that the existing signature returns a result
`fn ip_key(ip_str: &str) -> Result<String, Error>`

is because the IP address can be read from the `X-Real-Ip` header, which could be invalid.

by defaulting to 127.0.0.1 like this, if a proxy server is setup without the correct header then it would result in rate limiting silently failing to work correctly.

---

That is true but the existing uses unwrap when reading the ip. Unwrap should be avoided at all costs in production code because it crashes the entire server and is not easy for the average user to debug. Similarly returning an error in this place means that misconfiguration causes all incoming requests to fail. It is much better to ignore invalid IPs and apply the limit to localhost instead. Then it is easy to notice when running into rate limits too early that something is misconfigured in that area, without breaking the site completely. This is what weve been doing in production for years now without any complaints.